---
title: "VDJdb summary statistics"
author: "Mikhail Shugay"
date: "26-02-2022"
output:
  html_document:
    code_folding: hide
---

```{r setup, include=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# Make sure to keep the default for normal processing.
default_output_hook <- knitr::knit_hooks$get("output")

# Output hooks handle normal R console output.
knitr::knit_hooks$set( output = function(x, options) {

  comment <- knitr::opts_current$get("comment")
  if( is.na(comment) ) comment <- ""
  can_null <- grepl( paste0( comment, "\\s*\\[\\d?\\]" ),
                     x, perl = TRUE)
  do_null <- isTRUE( knitr::opts_current$get("null_prefix") )
  if( can_null && do_null ) {
    # By default R print output aligns at the right brace.
    align_index <- regexpr( "\\]", x )[1] - 1
    # Two cases: start or newline
    re <- paste0( "^.{", align_index, "}\\]")
    rep <- comment
    x <- gsub( re, rep,  x )
    re <- paste0( "\\\n.{", align_index, "}\\]")
    rep <- paste0( "\n", comment )
    x <- gsub( re, rep,  x )
  }

  default_output_hook( x, options )

})

knitr::opts_template$set("kill_prefix"=list(comment=NA, null_prefix=TRUE))
```

```{r message=FALSE, warning=FALSE}
library(knitr)
library(ggplot2)
library(RColorBrewer)
library(data.table)
library(forcats)
library(ggh4x)
library(ggalluvial)
library(circlize)
library(ggrepel)
library(tidyverse)
library(httr)
library(xml2)
library(stringr)
library(gridExtra)
library(maps)
library(scatterpie)
select = dplyr::select

df = fread("../database/vdjdb.slim.txt", header=T, sep="\t")
```

---

!summary_embed_start!

```{r, null_prefix=TRUE, warning=FALSE}
paste("Last updated on", format(Sys.time(), '%d %B, %Y'))
```

#### Record statistics by species and TCR chain

General statistics. Note that general statistics are computed using the 'slim' database version. Slim version, for example, lists the same TCR sequence found in several donors/studies only once and selects representative V/J for a given CDR3aa clonotype and the best score across all redundant records.

```{r warning=FALSE, message=FALSE}
df.sg = df %>% 
  group_by(species, gene) %>%
  summarize(records = length(complex.id), 
            paired.records = sum(ifelse(complex.id=="0", 0, 1)),
            epitopes = length(unique(antigen.epitope)),
            publications = length(unique(str_split_fixed(reference.id, ",", n = Inf)[,1]))) %>%
  arrange(species, gene)

colnames(df.sg) = c("Species", "Chain", "Records", "Paired records", "Unique epitopes", "Studies")

kable(format = "html", df.sg)
```

#### Record statistics by year

```{r message=FALSE, warning=FALSE, echo=FALSE, cache=FALSE}
# Load data 

dt.vdjdb = fread("../database/vdjdb_full.txt", sep = "\t", fill=T)

dt.vdjdb.s = dt.vdjdb %>%
  filter(species != "MacacaMulatta") %>%
  mutate(tcr_key = paste(v.alpha, j.alpha, cdr3.alpha, v.beta, j.beta, cdr3.beta),
         mhc_key = paste(mhc.a, mhc.b),
         chains = ifelse(cdr3.alpha != "", ifelse(cdr3.beta != "", "paired", "TRA"), "TRB")) %>%
  select(reference.id, tcr_key, mhc_key, chains, antigen.epitope, species) %>%
  unique

# Fetch pubmed info

ids = unique((dt.vdjdb.s %>%
  filter(substr(reference.id,1,4)=="PMID") %>%
  mutate(id = str_split_fixed(reference.id, ":", 2)[,2]))$id)

pm_data = content(GET("https://eutils.ncbi.nlm.nih.gov/", 
                      path = "entrez/eutils/esummary.fcgi",
                      query = list(db = "pubmed", 
                                   id = paste0(ids,collapse = ","))))

invisible(write_xml(pm_data, "dates.xml"))
pm_data_2 = readLines("dates.xml")
tmp=file.remove("dates.xml")

pm_data_id = str_split_fixed(pm_data_2[grepl("<Id>", pm_data_2)], "[<>]", n = Inf)[,3]
pm_data_date = str_split_fixed(pm_data_2[grepl('Name="PubDate"', pm_data_2)], "[<>]", n = Inf)[,3]

dt.pubdate = data.table(
  reference.id = paste0("PMID:", pm_data_id),
  pub_date = str_split_fixed(pm_data_date, " ", n = Inf)[,1]
)

dt.pubdate = rbind(dt.pubdate,
                   data.frame(reference.id = 
                                c("https://www.10xgenomics.com/resources/application-notes/a-new-way-of-exploring-immunity-linking-highly-multiplexed-antigen-recognition-to-immune-repertoire-and-phenotype/#",
                                  "http://mediatum.ub.tum.de/doc/1136748",
                                  "https://github.com/antigenomics/vdjdb-db/issues/288",
                                  "https://github.com/antigenomics/vdjdb-db/issues/193",
                                  "https://github.com/antigenomics/vdjdb-db/issues/252",
                                  "https://github.com/antigenomics/vdjdb-db/issues/313",
                                  "https://github.com/antigenomics/vdjdb-db/issues/312",
                                  "https://github.com/antigenomics/vdjdb-db/issues/311",
                                  "https://github.com/antigenomics/vdjdb-db/issues/315",
                                  "https://github.com/antigenomics/vdjdb-db/issues/323",
                                  "https://github.com/antigenomics/vdjdb-db/issues/326",
                                  "https://github.com/antigenomics/vdjdb-db/issues/328",
                                  "https://doi.org/10.1101/2020.05.04.20085779",
                                  "https://github.com/antigenomics/vdjdb-db/issues/333"),
                              pub_date = c(2019, 
                                           2014, 
                                           2019, 
                                           2017, 
                                           2017, 
                                           2020,
                                           2020,
                                           2020,
                                           2020,
                                           2021,
                                           2021,
                                           2021,
                                           2021,
                                           2021) %>% as.character, stringsAsFactors = F))
```

Number of unique TCR sequences, epitopes, MHC alleles and studies by **publication** year, cumulative plots.

```{r message=FALSE, warning=FALSE}
dt.vdjdb.s2 = dt.vdjdb.s %>%
  merge(dt.pubdate %>% mutate(pub_date = pub_date %>% as.integer), 
        by = "reference.id",
        allow.cartesian = T) %>%
  merge(expand.grid(pub_date = unique(dt.pubdate$pub_date) %>% as.integer,
                    pub_date2 = unique(dt.pubdate$pub_date) %>% as.integer), 
        by = "pub_date",
        allow.cartesian = T) %>%
  group_by(pub_date2, chains) %>%
  summarise(tcr_count = length(unique(tcr_key[which(pub_date <= pub_date2)])),
            epi_count = length(unique(antigen.epitope[which(pub_date <= pub_date2)])),
            ref_count = length(unique(reference.id[which(pub_date <= pub_date2)])),
            mhc_count = length(unique(mhc_key[which(pub_date <= pub_date2)])))

p1=ggplot(dt.vdjdb.s2, aes(x = as.integer(pub_date2), y = tcr_count, color = chains)) +
  geom_line() +
  geom_point() +
  ylab("") +
  scale_x_continuous("", breaks = seq(1995, 2024, by =2)) +
  scale_color_brewer(palette = "Set1") +
  theme_classic() + ggtitle("Number of unique TCRs") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5),
        axis.line = element_line(size = 0.3))

p2=ggplot(dt.vdjdb.s2, aes(x = as.integer(pub_date2), y = epi_count, color = chains)) +
  geom_line() +
  geom_point() +
  ylab("") +
  scale_x_continuous("", breaks = seq(1995, 2024, by =2)) +
  scale_color_brewer(palette = "Set1") +
  theme_classic() + ggtitle("Number of unique epitopes") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5),
        axis.line = element_line(size = 0.3))

p3=ggplot(dt.vdjdb.s2, aes(x = as.integer(pub_date2), y = ref_count, color = chains)) +
  geom_line() +
  geom_point() +
  ylab("") +
  scale_x_continuous("", breaks = seq(1995, 2024, by =2)) +
  scale_color_brewer(palette = "Set1") +
  theme_classic() + ggtitle("Number of studies") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5),
        axis.line = element_line(size = 0.3))

p4=ggplot(dt.vdjdb.s2, aes(x = as.integer(pub_date2), y = mhc_count, color = chains)) +
  geom_line() +
  geom_point() +
  ylab("") +
  scale_x_continuous("", breaks = seq(1995, 2024, by =2)) +
  scale_color_brewer("TCR chain(s)", palette = "Set1") +
  theme_classic() + ggtitle("Number of MHC alleles") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5),
        axis.line = element_line(size = 0.3)) +
  theme(legend.position="bottom")

g_legend = function(a.gplot){
  tmp <- ggplot_gtable(ggplot_build(a.gplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend)
  }

mylegend = g_legend(p4)

grid.arrange(arrangeGrob(p1 + theme(legend.position="none"),
                         p2 + theme(legend.position="none"),
                         p3 + theme(legend.position="none"),
                         p4 + theme(legend.position="none"),
                         nrow=2),
             mylegend, nrow=2,heights=c(10, 1)) #-> PX1

#PX1 %>% plot
#pdf("pubyear.pdf")
#PX1 %>% plot
#dev.off()

#fwrite(dt.vdjdb.s2, "vdjdb_stats_pubyear.txt", sep = "\t")
```

#### Summary by antigen and antigen origin

Representative data for Homo Sapiens

```{r message=FALSE, warning=FALSE}
df.a = df %>% 
  filter(species == "HomoSapiens") %>%
  group_by(antigen.species) %>% #, antigen.gene) %>%
  summarize(records = n(), 
            epitopes = length(unique(antigen.epitope)),
            publications = length(unique(str_split_fixed(reference.id, ",", n = Inf)[,1]))) %>%
  arrange(-records)

colnames(df.a) = c("Parent species", #"Parent gene",
                   "Records", "Unique epitopes", "Studies")
kable(format = "html", df.a)
```

---

#### **COVID-19** data

Summary of antigens and T-cell receptors related to COVID-19 pandemic. Number of records for SARS-CoV-2 epitopes grouped by viral protein and HLA plotted using alluvium plot. Epitopes with less than 30 records in total were not counted.

```{r message=FALSE, warning=FALSE, fig.width=12, fig.height=10}
df %>% 
  filter(species == "HomoSapiens", 
         startsWith(as.character(antigen.species), "SARS-CoV")) %>%
  mutate(mhc.a = str_split_fixed(mhc.a, "[,:]", 2)[,1]) %>%
  group_by(antigen.gene, mhc.a, antigen.epitope) %>%
  mutate(publications = length(unique(str_split_fixed(reference.id, ",", n = Inf)[,1]))) %>%
  group_by(antigen.gene, mhc.a, antigen.epitope, gene, publications) %>%
  summarize(records = n()) -> df.c

colnames(df.c) = c("Gene", "HLA", "Epitope", "TCR chain",
                   "Studies",
                   "Records")

ggplot(df.c %>% 
         ungroup %>% arrange(Records) %>% filter(Records >= 30),
       aes(axis1 = Gene,
           axis2 = gsub("HLA-", "", HLA),
           axis3 = substr(Epitope,1,3),
           axis4 = `TCR chain`,
           y = log2(Records))) +
  geom_alluvium(aes(fill = substr(Epitope,1,3) %>% as.factor %>% as.integer), 
                color = "white", alpha = 0.8, curve_type = "sigmoid") +
  geom_stratum(fill = "white", color = "black", size=0.2) +
  geom_text(stat = "stratum", aes(label = after_stat(stratum))) +
  scale_fill_distiller(palette = "Set3", guide=F, "") +
  #scale_fill_hue(guide=F, "") +
  ylab("") + scale_x_discrete(limits = c("Gene", "HLA", "Epitope", "TCR chain"),
                              expand = c(.1, .1),
                              position = "top") +
  theme_void() +
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.x =  element_text(size = 16, color = "black"),
        axis.ticks.x = element_blank(),
        panel.grid.major.y = element_blank(),
        legend.position = "bottom")
```

Summary of SARS-CoV-2 epitopes and corresponding TCR alpha and beta chain specificity records (cases with 10+ records)

```{r message=FALSE, warning=FALSE}
kable(format = "html", 
      df.c %>% reshape2::dcast(Gene + HLA + Epitope + Studies ~ `TCR chain`, fill = 0) %>%
        filter(TRA+TRB >= 10) %>%
        arrange(-(TRB+TRA)))
```

---

#### **Neoantigen** data

Summary of potential neoantigen targets for immunotherapy and T-cell receptors recognizing them. Number of records for neoantigens grouped by mutated gene and HLA plotted using alluvium plot. Neoantigens with less than 10 records in total were not counted.

```{r message=FALSE, warning=FALSE, fig.width=12, fig.height=10}
df %>% 
  filter(species == "HomoSapiens", 
         startsWith(as.character(antigen.species), "HomoSapiens")) %>%
  mutate(mhc.a = str_split_fixed(mhc.a, "[,:]", 2)[,1]) %>%
  group_by(antigen.gene, mhc.a, antigen.epitope) %>%
  mutate(publications = length(unique(str_split_fixed(reference.id, ",", n = Inf)[,1]))) %>%
  group_by(antigen.gene, mhc.a, antigen.epitope, gene, publications) %>%
  summarize(records = n()) -> df.n

colnames(df.n) = c("Gene", "HLA", "Epitope", "TCR chain",
                   "Studies",
                   "Records")

ggplot(df.n %>% 
         ungroup %>% arrange(Records) %>% filter(Records >= 10),
       aes(axis1 = Gene,
           axis2 = gsub("HLA-", "", HLA),
           axis3 = substr(Epitope,1,3),
           axis4 = `TCR chain`,
           y = log2(Records))) +
  geom_alluvium(aes(fill = substr(Epitope,1,3) %>% as.factor %>% as.integer), 
                color = "white", alpha = 0.8, curve_type = "sigmoid") +
  geom_stratum(fill = "white", color = "black", size=0.2) +
  geom_text(stat = "stratum", aes(label = after_stat(stratum))) +
  scale_fill_distiller(palette = "Accent", guide=F, "") +
  #scale_fill_hue(guide=F, "") +
  ylab("") + scale_x_discrete(limits = c("Gene", "HLA", "Epitope", "TCR chain"),
                              expand = c(.1, .1),
                              position = "top") +
  theme_void() +
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.x =  element_text(size = 16, color = "black"),
        axis.ticks.x = element_blank(),
        panel.grid.major.y = element_blank(),
        legend.position = "bottom")
```

Summary of neoantigens and corresponding TCR alpha and beta chain specificity records (cases with 5+ records)

```{r message=FALSE, warning=FALSE}
kable(format = "html", 
      df.n %>% reshape2::dcast(Gene + HLA + Epitope + Studies ~ `TCR chain`, fill = 0) %>%
        filter(TRA+TRB >= 5) %>%
        arrange(-(TRB+TRA)))
```

---

#### Distribution VDJdb confidence scores

Legend: 0 - low confidence, 1 - medium confidence, 2 - high confidence, 3 - very high confidence.

Note that this scoring system is currently deprecated due to large amounts of deep multimer+ repertoire sequencing data and 10X scRNAseq with ImmuDEX dextramer multiplex that are almost impossible to validate independently. We suggest applying methods like TCRNET to [infer high-confidence TCR motifs](https://github.com/antigenomics/vdjdb-motifs) from VDJdb.

```{r message=FALSE, warning=FALSE}
df.score <- df %>%
  group_by(mhc.class, gene, vdjdb.score) %>%
  summarize(total = n())

ggplot(df.score, aes(x=paste(mhc.class, gene, sep = " "), y=total, fill = as.factor(vdjdb.score))) + 
  geom_bar(stat = "identity", position = "dodge", color = "black", size = 0.3) +  
  xlab("") + scale_y_log10("Records") +
  scale_fill_brewer("VDJdb score", palette = "PuBuGn") + 
  theme_classic() +
  theme(legend.position="bottom",
        axis.line = element_line(size = 0.3))
```

---

#### Spectratype

Representative data for Homo Sapiens. The spectratype is colored by epitope.

```{r message=FALSE, warning=FALSE}
df.spe = subset(df, species=="HomoSapiens")

ggplot(df.spe %>% mutate(
  epi_len = nchar(antigen.epitope),
  antigen.epitope = as.factor(antigen.epitope)), 
  aes(x=nchar(as.character(cdr3)))) + 
  geom_histogram(aes(fill = antigen.epitope %>% fct_reorder(epi_len) %>% as.integer(), 
                     group = antigen.epitope), #), 
                 bins = 21, size = 0,
                 binwidth = 1, alpha = 0.9, color = NA) + 
  geom_density(aes(y=..count..), adjust = 4, linetype = "dotted") +
  scale_x_continuous(limits = c(5,25), breaks = seq(5,25,5)) + 
  facet_wrap(~gene) + 
  scale_fill_distiller(palette = "Spectral", guide=F, "") +
  #scale_fill_viridis_d(guide=F, direction = -1) +
  xlab("CDR3 length") + ylab("Records") +
  theme_classic() +
  theme(axis.line = element_line(size = 0.3),
        strip.background = element_blank())
```

#### V(D)J usage and MHC alleles

Representative data for Homo Sapiens, Variable gene. Only Variable genes and MHC alleles with at least 10 records are shown.

```{r message=FALSE, warning=FALSE, fig.height=10, fig.width=6}
df.vhla = df %>% 
  as.tibble %>%
  filter(species == "HomoSapiens" ) %>%
  mutate(id = 1:n()) %>%
  separate_rows(mhc.a, sep = ",") %>%
  separate_rows(mhc.b, sep = ",") %>%
  separate_rows(v.segm, sep = ",") %>%
  mutate(mhc.a.split = str_split_fixed(mhc.a, fixed(":"), n = Inf)[,1],
         mhc.b.split = str_split_fixed(mhc.b, fixed(":"), n = Inf)[,1],
         v.segm.split = str_split_fixed(v.segm, fixed("*"), n = Inf)[,1]) %>%
  group_by(gene, mhc.class, mhc.a.split, mhc.b.split, v.segm.split) %>%
  summarize(records = length(unique(id))) %>%
  group_by(mhc.class, mhc.a.split, mhc.b.split) %>%
  mutate(records.mhc = sum(records)) %>%
  group_by(gene, v.segm.split) %>%
  mutate(records.v = sum(records)) %>%
  filter(records.mhc >= 10, records.v >= 10)
  
#df.vhla$v.segm.split = with(df.vhla, factor(v.segm.split, v.segm.split[order(records.v)]))
#df.vhla$mhc.a.split = with(df.vhla, factor(mhc.a.split, mhc.a.split[order(records.mhc)]))

ggplot(df.vhla, aes(x=gsub("HLA-", "", paste(mhc.a.split, mhc.b.split, sep = " / ")) %>%
                      fct_reorder(records), 
                    y=v.segm.split %>%
                      fct_reorder(records), fill = pmin(records, 1000))) +
  geom_tile() +
  scale_fill_gradientn("Records", colors=colorRampPalette(brewer.pal(9, 'PuBuGn'))(32), 
                       trans="log", breaks = c(1, 10, 100, 1000)) +
  xlab("") + ylab("") +
  facet_grid(gene~mhc.class, scales="free", space="free") +
  theme_classic() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.text.y = element_text(size = 8),
        panel.grid.major = element_blank(),
        legend.position = "right",
        axis.line = element_line(size = 0.3),
        strip.background = element_blank())
```

Circos plot for correspondence between human TRBV genes and MHC class I alleles for links supported by more than 50 records in terms of the ratio of observed to expected records. Band width and color (from red:highest to light yellow:lowest) are scaled proportional to co-occurrence matrix value divided by row (TRBV) and column (HLA) sums.

```{r message=FALSE, warning=FALSE, fig.width=8, fig.height=8}
#https://jokergoo.github.io/circlize_book
df.vhla.1b <- df.vhla %>%
  filter(records >= 50) %>%
  filter(gene == "TRB", mhc.class == "MHCI") %>%
  mutate(mhc.a.split = substr(mhc.a.split, 5, nchar(mhc.a.split)),
         v.segm.split = paste0("Vb", substr(v.segm.split, 5, nchar(v.segm.split)))) %>%
  reshape2::dcast(mhc.a.split ~ v.segm.split, value.var = "records", fill = 0)
rownames(df.vhla.1b) <- df.vhla.1b$mhc.a.split
df.vhla.1b$mhc.a.split <- NULL
df.vhla.1b <- df.vhla.1b %>% as.matrix()
df.vhla.1b <- (df.vhla.1b %>% 
  sweep(1, rowSums(df.vhla.1b), `/`) %>%
  sweep(2, colSums(df.vhla.1b), `/`)) * sum(df.vhla.1b)

grid.col <- setNames(c(colorRampPalette(brewer.pal(9, 'YlGn'))(nrow(df.vhla.1b)) %>% rev, 
                       colorRampPalette(brewer.pal(9, 'PuBu'))(ncol(df.vhla.1b))), 
                     union(rownames(df.vhla.1b), colnames(df.vhla.1b)))
band.col <- colorRamp2(range(df.vhla.1b), 
                       colorRampPalette(brewer.pal(9, 'YlOrRd'))(2), 
                       transparency = 0.2)

chordDiagram(df.vhla.1b, annotationTrack = "grid", preAllocateTracks = 1, 
             col = band.col,
             grid.col = grid.col,
             grid.border = "black",
             big.gap = 20)

circos.trackPlotRegion(track.index = 1, panel.fun = function(x, y) {
  xlim = get.cell.meta.data("xlim")
  ylim = get.cell.meta.data("ylim")
  sector.name = get.cell.meta.data("sector.index")
  circos.text(mean(xlim), ylim[1] + .1, sector.name, 
              facing = "clockwise", niceFacing = T, adj = c(0, 0.5), 
              cex = 0.65)
}, bg.border = NA)

title("TRBV ~ HLA-I")

circos.clear()
```

---

#### Detailed summary for HLA

Representative data for Homo Sapiens MHC class I and II

```{r message=FALSE, warning=FALSE}
df.m = df %>% 
  as.tibble %>%
  filter(species == "HomoSapiens") %>%
  mutate(id = 1:n()) %>%
  separate_rows(mhc.a, sep = ",") %>%
  separate_rows(mhc.b, sep = ",") %>%
  mutate(mhc.a.split = str_split_fixed(mhc.a, fixed(":"), n = Inf)[,1],
         mhc.b.split = str_split_fixed(mhc.b, fixed(":"), n = Inf)[,1]) %>%
  group_by(mhc.a.split, mhc.b.split) %>%
  summarize(records = length(unique(id)), 
            antigens = length(unique(antigen.epitope)),
            publications = length(unique(str_split_fixed(reference.id, ",", n = Inf)[,1]))) %>%
  arrange(-records)
  
colnames(df.m) = c("First chain", "Second chain", "Records", "Unique epitopes", "Studies")
kable(format = "html", df.m)
```

!summary_embed_end!

---

#### VDJdb 2022 paper

Total COVID records

```{r}
df %>%
  filter(species == "HomoSapiens", 
         startsWith(as.character(antigen.species), "SARS-CoV")) %>%
  nrow()
```

Total COVID epitopes

```{r}
df %>%
  filter(species == "HomoSapiens", 
         startsWith(as.character(antigen.species), "SARS-CoV")) %>%
  .$antigen.epitope %>%
  unique %>% 
  length
```

According to VDJdb'19 "since the establishment of the database a total of 155 published studies were processed and added resulting in 61049 TCR specificity records in 2019 compared to only 5491 in 2017."

Count total new records

```{r}
nrow(df) - 61049
```

Total number of publications

```{r}
dt.pubdate$reference.id %>%
  unique %>%
  length
```

Number of epitopes ("Notably, while the number of unique epitopes with known specific TCRs increased, it is still dwarfed by the number of unique TCRs in the database, as there are currently only 212 epitopes in VDJdb. " for 2019)

```{r}
df %>%
  .$antigen.epitope %>%
  unique %>% 
  length
```

Numbers of records and epitopes for common viruses.

CMV

```{r}
df %>% filter(antigen.species == "CMV") %>% nrow
df %>% filter(antigen.species == "CMV") %>% .$antigen.epitope %>% unique %>% length
```

Flu

```{r}
df %>% filter(antigen.species == "InfluenzaA") %>% nrow
df %>% filter(antigen.species == "InfluenzaA") %>% .$antigen.epitope %>% unique %>% length
```

EBV

```{r}
df %>% filter(antigen.species == "EBV") %>% nrow
df %>% filter(antigen.species == "EBV") %>% .$antigen.epitope %>% unique %>% length
```

Count number of HLA haplotypes. Class I

```{r}
df %>%
  filter(species == "HomoSapiens", mhc.class == "MHCI") %>%
  mutate(allele = str_split_fixed(mhc.a, "[:,]", 2)[,1]) %>%
  .$allele %>%
  unique %>%
  length
```

class II

```{r}
df %>%
  filter(species == "HomoSapiens", mhc.class == "MHCII") %>%
  mutate(allele1 = str_split_fixed(mhc.a, "[:,]", 2)[,1],
         allele2 = str_split_fixed(mhc.b, "[:,]", 2)[,1],
         allele = paste(allele1, allele2)) %>%
  .$allele %>%
  unique %>%
  length
```

Count number of records for HLA-A*02

```{r}
df %>%
  mutate(allele = str_split_fixed(mhc.a, "[:,]", 2)[,1]) %>%
  filter(allele == "HLA-A*02") %>%
  nrow()
```

and HLA-A*03

```{r}
df %>%
  mutate(allele = str_split_fixed(mhc.a, "[:,]", 2)[,1]) %>%
  filter(allele == "HLA-A*03") %>%
  nrow()
```

Number of studies with self-antigens

```{r}
df %>%
  filter(antigen.species == "HomoSapiens") %>%
  .$reference.id %>%
  unique %>%
  length
```
number of records in them

```{r}
df %>%
  filter(antigen.species == "HomoSapiens") %>%
  nrow()
```

and the number of unique epitopes

```{r}
df %>%
  filter(antigen.species == "HomoSapiens") %>%
  .$antigen.epitope %>%
  unique %>%
  length
```

COVID19 - number of records in original Shomuradova paper

```{r}
df %>%
  filter(reference.id == "https://github.com/antigenomics/vdjdb-db/issues/315") %>%
  nrow
```

total records by 2022

```{r}
df %>%
  filter(antigen.species == "SARS-CoV-2") %>%
  nrow
```

and number of epitopes

```{r}
df %>%
  filter(antigen.species == "SARS-CoV-2") %>%
  .$antigen.epitope %>%
  unique %>%
  length
```

Figure1b and TableS1

```{r}
pdf("covid_alluvium.pdf", width = 12, height = 10)
p1
dev.off()

df.c %>% reshape2::dcast(Gene + HLA + Epitope + Studies ~ `TCR chain`, fill = 0) %>%
        arrange(-(TRB+TRA)) %>%
  fwrite("covid_summary.txt", sep = "\t")
```

```{r fig.width=9, fig.height=4.5}
df %>%
  filter(antigen.species == "SARS-CoV-2") %>%
  select(antigen.epitope, reference.id) %>%
  separate_rows(reference.id, sep=",") %>%
  group_by(reference.id, antigen.epitope) %>%
  summarise(count = n()) %>%
  group_by(reference.id) %>%
  mutate(total = sum(count)) -> cov19.by.lab

#unique(cov19.by.lab$reference.id)

cov19.by.lab.coords <- tibble(
  reference.id = c(
    "https://doi.org/10.1101/2020.05.04.20085779",        
    "https://github.com/antigenomics/vdjdb-db/issues/315",
    "https://github.com/antigenomics/vdjdb-db/issues/323",
    "https://github.com/antigenomics/vdjdb-db/issues/326",
    "https://github.com/antigenomics/vdjdb-db/issues/333",
    "PMID:33664060",                                      
    "PMID:33951417",                                      
    "PMID:34061349",                                      
    "PMID:34086357",                                      
    "PMID:34222571"),
  city = c("Seattle",
           "Moscow",
           "Cardiff",
           "Memphis",
           "Moscow",
           "Vienna",
           "Melbourne",
           "Erlangen",
           "Melbourne",
           "Chongqing"),
  long = c(
    -122.3351,        
    37.6184,
    -3.1790,
    -89.9711,
    37.6184,
    16.3634, 
    144.9464,                                      
    11.0166,                                      
    144.9464,                                      
    106.8877),
  lat = c(
    47.6080,        
    55.7512,
    51.4815,
    35.1175,
    55.7512,
    48.2100,
    -37.8409,                                      
    49.5833,                                      
    -37.8409,                                      
    29.4393))

cov19.by.lab <- cov19.by.lab %>%
  merge(unique(cov19.by.lab.coords))

cov19.by.lab.pie <- cov19.by.lab %>%
                    reshape2::dcast(city + long + lat + total ~ antigen.epitope, 
                                    value.var = "count", fill = 0)

world <- map_data("world")

set.seed(21)
ggplot() +
  geom_map(
    data = world, map = world,
    aes(long, lat, map_id = region),
    color = "white", 
    fill = "grey",
    size = 0.2
  ) +
  #geom_point(data = cov19.by.lab, aes(long, lat, size = total)) +
  geom_scatterpie(data = cov19.by.lab.pie, 
                  aes(x=long, y=lat, r = log(total + 1)),
                  colnames(cov19.by.lab.pie)[5:ncol(cov19.by.lab.pie)],
                  color = NA) +
  geom_text_repel(data = cov19.by.lab %>% select(long, lat, city) %>% unique, 
            aes(long, lat, label = city),
            alpha = 0.8) +
  scale_fill_hue(guide=F, "") +
  scale_y_continuous(limits=c(-55, 85)) +
  scale_x_continuous(limits=c(-150, 170)) +
  coord_quickmap() +
  geom_scatterpie_legend(log(cov19.by.lab.pie$total + 1), 
                         x=-125, y=-30, n=3, labeller=function(x) round(exp(x) - 1, 0)) +
  theme_void() +
  theme(legend.position = "none") -> p0

p0

pdf("covid_geo.pdf")
p0
dev.off()
```

---

#### Misc

Number of unique CDR3 sequences per epitope

```{r warning=FALSE, message=FALSE, fig.height=6, fig.width=6}
df.ag = df %>%
  filter(species == "HomoSapiens") %>%
  group_by(gene, mhc.class, antigen.species, antigen.gene, antigen.epitope) %>%
  summarize(cdr3.count = length(unique(cdr3))) %>%
  arrange(-cdr3.count)

ggplot(df.ag, aes(x = cdr3.count, y = rank(-cdr3.count))) +
  #geom_histogram() +
  geom_smooth(method = "lm", 
              color = "red", 
              size = 0.5) +
  geom_point(color = "#253494", alpha = 0.3) +
  scale_x_log10("Unique CDR3s") + scale_y_log10("Epitope rank") +
  facet_grid(mhc.class~gene) +
  theme_classic()
```

Tabular

```{r warning=FALSE, message=FALSE}
colnames(df.ag) = c("TCR chain", "MHC class", "Parent species", "Parent gene", "Epitope sequence", "Unique CDR3")
kable(format = "html", df.ag)
```

Epitopes, parent genes and species - flat table for checking naming

```{r warning=FALSE, message=FALSE}
df %>%
  filter(species == "HomoSapiens") %>%
  select(antigen.species, antigen.gene, antigen.epitope) %>%
  unique %>%
  group_by(antigen.epitope) %>%
  mutate(ambiguous.naming = ifelse(n() > 1, "*", "")) %>%
  arrange(antigen.species, antigen.gene, antigen.epitope) %>%
  kable(format = "html")
```

```{r}
#END
```