{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ce00d389",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d511a168",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>length</th>\n",
       "      <th>chain</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>c_gene</th>\n",
       "      <th>...</th>\n",
       "      <th>cdr3_nt</th>\n",
       "      <th>fwr4</th>\n",
       "      <th>fwr4_nt</th>\n",
       "      <th>reads</th>\n",
       "      <th>umis</th>\n",
       "      <th>raw_clonotype_id</th>\n",
       "      <th>raw_consensus_id</th>\n",
       "      <th>exact_subclonotype_id</th>\n",
       "      <th>donor</th>\n",
       "      <th>origin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGCAGCCACCA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGCAGCCACCA-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>577</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV12-2+TRBV13-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-7</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>...</td>\n",
       "      <td>TGTGCCAGCGGTGATGCCCTGGGACTGGGGGTCTCCTATGAACAGT...</td>\n",
       "      <td>GPGTRLTVL</td>\n",
       "      <td>GGTCCCGGCACCAGGCTCACGGTTTTAG</td>\n",
       "      <td>12284</td>\n",
       "      <td>9</td>\n",
       "      <td>clonotype8232</td>\n",
       "      <td>clonotype8232_consensus_1</td>\n",
       "      <td>1</td>\n",
       "      <td>Donor-2</td>\n",
       "      <td>TC-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGCATTTCAGG-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGCATTTCAGG-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>664</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV5-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ42</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>...</td>\n",
       "      <td>TGTGCTGCAAGTGTGGCTGGAGGAAGCAATGCAAAGCTAACCTTC</td>\n",
       "      <td>GKGTKLSVKS</td>\n",
       "      <td>GGGAAAGGCACTAAACTCTCTGTTAAATCAA</td>\n",
       "      <td>3968</td>\n",
       "      <td>4</td>\n",
       "      <td>clonotype1</td>\n",
       "      <td>clonotype1_consensus_2</td>\n",
       "      <td>1</td>\n",
       "      <td>Donor-2</td>\n",
       "      <td>TC-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGCATTTCAGG-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGCATTTCAGG-1_contig_2</td>\n",
       "      <td>True</td>\n",
       "      <td>661</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV13-3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-7</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>...</td>\n",
       "      <td>TGTGCCAGCAGTGAAGCCGGGACTGGGGGATATGAACAGTACTTC</td>\n",
       "      <td>GPGTRLTVL</td>\n",
       "      <td>GGTCCCGGCACCAGGCTCACGGTTTTAG</td>\n",
       "      <td>4818</td>\n",
       "      <td>4</td>\n",
       "      <td>clonotype1</td>\n",
       "      <td>clonotype1_consensus_1</td>\n",
       "      <td>1</td>\n",
       "      <td>Donor-2</td>\n",
       "      <td>TC-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACGGGAGGACTGGT-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACGGGAGGACTGGT-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>617</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV5-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ50</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>...</td>\n",
       "      <td>TGTGCTGCAAGCTTAGCATCCTCCTCCTTCAGCAAGCTGGTGTTT</td>\n",
       "      <td>GQGTSLSVVP</td>\n",
       "      <td>GGGCAGGGGACATCCTTATCAGTCGTTCCAA</td>\n",
       "      <td>2858</td>\n",
       "      <td>3</td>\n",
       "      <td>clonotype5975</td>\n",
       "      <td>clonotype5975_consensus_1</td>\n",
       "      <td>1</td>\n",
       "      <td>Donor-2</td>\n",
       "      <td>TC-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGGGAGTGCCATT-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACGGGAGTGCCATT-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>567</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV12N-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ31</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>...</td>\n",
       "      <td>TGTGCTCTGAATAACAATAACAGAATCTTCTTT</td>\n",
       "      <td>GDGTQLVVKP</td>\n",
       "      <td>GGTGATGGGACGCAGCTGGTGGTGAAGCCCA</td>\n",
       "      <td>9288</td>\n",
       "      <td>6</td>\n",
       "      <td>clonotype7</td>\n",
       "      <td>clonotype7_consensus_1</td>\n",
       "      <td>1</td>\n",
       "      <td>Donor-2</td>\n",
       "      <td>TC-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15482</th>\n",
       "      <td>TTTGTCACAAGTACCT-6</td>\n",
       "      <td>True</td>\n",
       "      <td>TTTGTCACAAGTACCT-6_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>595</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV5-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ34</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>...</td>\n",
       "      <td>TGTGCTGCCGGATCTTCCAATACCGACAAAGTCGTCTTT</td>\n",
       "      <td>GTGTRLQVSP</td>\n",
       "      <td>GGAACAGGGACCAGATTACAAGTCTCACCAA</td>\n",
       "      <td>18310</td>\n",
       "      <td>5</td>\n",
       "      <td>clonotype3944</td>\n",
       "      <td>clonotype3944_consensus_1</td>\n",
       "      <td>1</td>\n",
       "      <td>Donor-1</td>\n",
       "      <td>TC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15483</th>\n",
       "      <td>TTTGTCATCAAACCAC-6</td>\n",
       "      <td>True</td>\n",
       "      <td>TTTGTCATCAAACCAC-6_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>580</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV12N-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ22</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>...</td>\n",
       "      <td>TGTGCTCTGAGTGAATCTTCTGGCAGCTGGCAACTCATCTTT</td>\n",
       "      <td>GSGTQLTVMP</td>\n",
       "      <td>GGATCTGGAACCCAACTGACAGTTATGCCTG</td>\n",
       "      <td>14390</td>\n",
       "      <td>6</td>\n",
       "      <td>clonotype26</td>\n",
       "      <td>clonotype26_consensus_1</td>\n",
       "      <td>1</td>\n",
       "      <td>Donor-1</td>\n",
       "      <td>TC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15484</th>\n",
       "      <td>TTTGTCATCACTCTTA-6</td>\n",
       "      <td>True</td>\n",
       "      <td>TTTGTCATCACTCTTA-6_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>599</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV5-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ50</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>...</td>\n",
       "      <td>TGTGCTGTGTCATCCTCCTCCTTCAGCAAGCTGGTGTTT</td>\n",
       "      <td>GQGTSLSVVP</td>\n",
       "      <td>GGGCAGGGGACATCCTTATCAGTCGTTCCAA</td>\n",
       "      <td>11416</td>\n",
       "      <td>4</td>\n",
       "      <td>clonotype3963</td>\n",
       "      <td>clonotype3963_consensus_1</td>\n",
       "      <td>1</td>\n",
       "      <td>Donor-1</td>\n",
       "      <td>TC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15485</th>\n",
       "      <td>TTTGTCATCCGTAGTA-6</td>\n",
       "      <td>True</td>\n",
       "      <td>TTTGTCATCCGTAGTA-6_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>695</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-4</td>\n",
       "      <td>TRBC1</td>\n",
       "      <td>...</td>\n",
       "      <td>TGTGCCAGCAGTATGGCGCGGGGCGAAAGATTATTTTTC</td>\n",
       "      <td>GHGTKLSVL</td>\n",
       "      <td>GGTCATGGAACCAAGCTGTCTGTCCTGG</td>\n",
       "      <td>18316</td>\n",
       "      <td>7</td>\n",
       "      <td>clonotype5390</td>\n",
       "      <td>clonotype5390_consensus_1</td>\n",
       "      <td>1</td>\n",
       "      <td>Donor-1</td>\n",
       "      <td>TC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15486</th>\n",
       "      <td>TTTGTCATCGGGAGTA-6</td>\n",
       "      <td>True</td>\n",
       "      <td>TTTGTCATCGGGAGTA-6_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>519</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV6N-7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ40</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>...</td>\n",
       "      <td>TGCGCTCTGAGTGGCCTTAATACAGGAAACTACAAATACGTCTTT</td>\n",
       "      <td>GAGTRLKVIA</td>\n",
       "      <td>GGAGCAGGTACCAGACTGAAGGTTATAGCAC</td>\n",
       "      <td>22832</td>\n",
       "      <td>5</td>\n",
       "      <td>clonotype7096</td>\n",
       "      <td>clonotype7096_consensus_1</td>\n",
       "      <td>1</td>\n",
       "      <td>Donor-1</td>\n",
       "      <td>TC-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15487 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  barcode  is_cell                    contig_id  \\\n",
       "0      AAACCTGCAGCCACCA-1     True  AAACCTGCAGCCACCA-1_contig_1   \n",
       "1      AAACCTGCATTTCAGG-1     True  AAACCTGCATTTCAGG-1_contig_1   \n",
       "2      AAACCTGCATTTCAGG-1     True  AAACCTGCATTTCAGG-1_contig_2   \n",
       "3      AAACGGGAGGACTGGT-1     True  AAACGGGAGGACTGGT-1_contig_1   \n",
       "4      AAACGGGAGTGCCATT-1     True  AAACGGGAGTGCCATT-1_contig_1   \n",
       "...                   ...      ...                          ...   \n",
       "15482  TTTGTCACAAGTACCT-6     True  TTTGTCACAAGTACCT-6_contig_1   \n",
       "15483  TTTGTCATCAAACCAC-6     True  TTTGTCATCAAACCAC-6_contig_1   \n",
       "15484  TTTGTCATCACTCTTA-6     True  TTTGTCATCACTCTTA-6_contig_1   \n",
       "15485  TTTGTCATCCGTAGTA-6     True  TTTGTCATCCGTAGTA-6_contig_1   \n",
       "15486  TTTGTCATCGGGAGTA-6     True  TTTGTCATCGGGAGTA-6_contig_1   \n",
       "\n",
       "       high_confidence  length chain             v_gene d_gene   j_gene  \\\n",
       "0                 True     577   TRB  TRBV12-2+TRBV13-2    NaN  TRBJ2-7   \n",
       "1                 True     664   TRA            TRAV5-4    NaN   TRAJ42   \n",
       "2                 True     661   TRB           TRBV13-3    NaN  TRBJ2-7   \n",
       "3                 True     617   TRA            TRAV5-4    NaN   TRAJ50   \n",
       "4                 True     567   TRA          TRAV12N-2    NaN   TRAJ31   \n",
       "...                ...     ...   ...                ...    ...      ...   \n",
       "15482             True     595   TRA            TRAV5-4    NaN   TRAJ34   \n",
       "15483             True     580   TRA          TRAV12N-2    NaN   TRAJ22   \n",
       "15484             True     599   TRA            TRAV5-4    NaN   TRAJ50   \n",
       "15485             True     695   TRB             TRBV19    NaN  TRBJ1-4   \n",
       "15486             True     519   TRA           TRAV6N-7    NaN   TRAJ40   \n",
       "\n",
       "      c_gene  ...                                            cdr3_nt  \\\n",
       "0      TRBC2  ...  TGTGCCAGCGGTGATGCCCTGGGACTGGGGGTCTCCTATGAACAGT...   \n",
       "1       TRAC  ...      TGTGCTGCAAGTGTGGCTGGAGGAAGCAATGCAAAGCTAACCTTC   \n",
       "2      TRBC2  ...      TGTGCCAGCAGTGAAGCCGGGACTGGGGGATATGAACAGTACTTC   \n",
       "3       TRAC  ...      TGTGCTGCAAGCTTAGCATCCTCCTCCTTCAGCAAGCTGGTGTTT   \n",
       "4       TRAC  ...                  TGTGCTCTGAATAACAATAACAGAATCTTCTTT   \n",
       "...      ...  ...                                                ...   \n",
       "15482   TRAC  ...            TGTGCTGCCGGATCTTCCAATACCGACAAAGTCGTCTTT   \n",
       "15483   TRAC  ...         TGTGCTCTGAGTGAATCTTCTGGCAGCTGGCAACTCATCTTT   \n",
       "15484   TRAC  ...            TGTGCTGTGTCATCCTCCTCCTTCAGCAAGCTGGTGTTT   \n",
       "15485  TRBC1  ...            TGTGCCAGCAGTATGGCGCGGGGCGAAAGATTATTTTTC   \n",
       "15486   TRAC  ...      TGCGCTCTGAGTGGCCTTAATACAGGAAACTACAAATACGTCTTT   \n",
       "\n",
       "             fwr4                          fwr4_nt  reads umis  \\\n",
       "0       GPGTRLTVL     GGTCCCGGCACCAGGCTCACGGTTTTAG  12284    9   \n",
       "1      GKGTKLSVKS  GGGAAAGGCACTAAACTCTCTGTTAAATCAA   3968    4   \n",
       "2       GPGTRLTVL     GGTCCCGGCACCAGGCTCACGGTTTTAG   4818    4   \n",
       "3      GQGTSLSVVP  GGGCAGGGGACATCCTTATCAGTCGTTCCAA   2858    3   \n",
       "4      GDGTQLVVKP  GGTGATGGGACGCAGCTGGTGGTGAAGCCCA   9288    6   \n",
       "...           ...                              ...    ...  ...   \n",
       "15482  GTGTRLQVSP  GGAACAGGGACCAGATTACAAGTCTCACCAA  18310    5   \n",
       "15483  GSGTQLTVMP  GGATCTGGAACCCAACTGACAGTTATGCCTG  14390    6   \n",
       "15484  GQGTSLSVVP  GGGCAGGGGACATCCTTATCAGTCGTTCCAA  11416    4   \n",
       "15485   GHGTKLSVL     GGTCATGGAACCAAGCTGTCTGTCCTGG  18316    7   \n",
       "15486  GAGTRLKVIA  GGAGCAGGTACCAGACTGAAGGTTATAGCAC  22832    5   \n",
       "\n",
       "      raw_clonotype_id           raw_consensus_id exact_subclonotype_id  \\\n",
       "0        clonotype8232  clonotype8232_consensus_1                     1   \n",
       "1           clonotype1     clonotype1_consensus_2                     1   \n",
       "2           clonotype1     clonotype1_consensus_1                     1   \n",
       "3        clonotype5975  clonotype5975_consensus_1                     1   \n",
       "4           clonotype7     clonotype7_consensus_1                     1   \n",
       "...                ...                        ...                   ...   \n",
       "15482    clonotype3944  clonotype3944_consensus_1                     1   \n",
       "15483      clonotype26    clonotype26_consensus_1                     1   \n",
       "15484    clonotype3963  clonotype3963_consensus_1                     1   \n",
       "15485    clonotype5390  clonotype5390_consensus_1                     1   \n",
       "15486    clonotype7096  clonotype7096_consensus_1                     1   \n",
       "\n",
       "         donor origin  \n",
       "0      Donor-2   TC-2  \n",
       "1      Donor-2   TC-2  \n",
       "2      Donor-2   TC-2  \n",
       "3      Donor-2   TC-2  \n",
       "4      Donor-2   TC-2  \n",
       "...        ...    ...  \n",
       "15482  Donor-1   TC-1  \n",
       "15483  Donor-1   TC-1  \n",
       "15484  Donor-1   TC-1  \n",
       "15485  Donor-1   TC-1  \n",
       "15486  Donor-1   TC-1  \n",
       "\n",
       "[15487 rows x 33 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv('GSE309696_filtered_contig_annotations.csv.gz')\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7834abdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "barcodes_dict = {}\n",
    "chains = ['TRA', 'TRB']\n",
    "\n",
    "\n",
    "for barcode in data.barcode.unique():\n",
    "    \n",
    "    barcodes_dict[barcode] = {}\n",
    "    for chain in chains:\n",
    "        data_i = data[(data.barcode == barcode) & (data.chain == chain)]\n",
    "        barcodes_dict[barcode][chain] = [\n",
    "            data_i.v_gene.to_list(),\n",
    "            data_i.d_gene.to_list(),            \n",
    "            data_i.j_gene.to_list(),\n",
    "            data_i.cdr3.to_list(),\n",
    "\n",
    "        ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e0c76367",
   "metadata": {},
   "outputs": [],
   "source": [
    "peptide_dict = {\n",
    "    'TC-1': np.nan,\n",
    "    'TC-2': np.nan,\n",
    "    'TP-1': 'VEALYLVCG',\n",
    "    'TP-2': 'VEALYLVCG',\n",
    "    'TS-1': 'VEALYLVSG',\n",
    "    'TS-2': 'VEALYLVSG'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "649ad1c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "data['antigen.epitope'] = data.origin.map(peptide_dict)\n",
    "\n",
    "data['antigen.gene'] = 'INS'\n",
    "data['antigen.species'] = 'HomoSapiens'\n",
    "data['reference.id'] = 'PMID:41315082'\n",
    "data['mhc.class'] = 'MHCII'\n",
    "data['mhc.a'] = 'HLA-DQA1*03:01'\n",
    "data['mhc.b'] = 'HLA-DQB1*03:02'\n",
    "\n",
    "data['species'] = 'MusMusculus'\n",
    "\n",
    "\n",
    "data['method.identification'] = 'tetramer-sort'\n",
    "data['method.singlecell'] = 'yes'\n",
    "data['method.sequencing'] = 'amplicon-seq'\n",
    "\n",
    "data['meta.tissue'] = 'spleen and lymph node'\n",
    "data['meta.cell.subset'] = 'CD4+'\n",
    "\n",
    "\n",
    "data['chunk.id'] = list(range(1, len(data)+1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0fb9d81d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>length</th>\n",
       "      <th>chain</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>c_gene</th>\n",
       "      <th>...</th>\n",
       "      <th>meta.tissue</th>\n",
       "      <th>meta.cell.subset</th>\n",
       "      <th>chunk.id</th>\n",
       "      <th>cdr3.alpha</th>\n",
       "      <th>v.alpha</th>\n",
       "      <th>j.alpha</th>\n",
       "      <th>cdr3.beta</th>\n",
       "      <th>v.beta</th>\n",
       "      <th>d.beta</th>\n",
       "      <th>j.beta</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGCAGCCACCA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGCAGCCACCA-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>577</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV12-2+TRBV13-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-7</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>...</td>\n",
       "      <td>spleen and lymph node</td>\n",
       "      <td>CD4+</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASGDALGLGVSYEQYF</td>\n",
       "      <td>TRBV12-2+TRBV13-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGCATTTCAGG-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGCATTTCAGG-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>664</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV5-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ42</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>...</td>\n",
       "      <td>spleen and lymph node</td>\n",
       "      <td>CD4+</td>\n",
       "      <td>2</td>\n",
       "      <td>CAASVAGGSNAKLTF</td>\n",
       "      <td>TRAV5-4</td>\n",
       "      <td>TRAJ42</td>\n",
       "      <td>CASSEAGTGGYEQYF</td>\n",
       "      <td>TRBV13-3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGCATTTCAGG-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGCATTTCAGG-1_contig_2</td>\n",
       "      <td>True</td>\n",
       "      <td>661</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV13-3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-7</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>...</td>\n",
       "      <td>spleen and lymph node</td>\n",
       "      <td>CD4+</td>\n",
       "      <td>3</td>\n",
       "      <td>CAASVAGGSNAKLTF</td>\n",
       "      <td>TRAV5-4</td>\n",
       "      <td>TRAJ42</td>\n",
       "      <td>CASSEAGTGGYEQYF</td>\n",
       "      <td>TRBV13-3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACGGGAGGACTGGT-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACGGGAGGACTGGT-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>617</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV5-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ50</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>...</td>\n",
       "      <td>spleen and lymph node</td>\n",
       "      <td>CD4+</td>\n",
       "      <td>4</td>\n",
       "      <td>CAASLASSSFSKLVF</td>\n",
       "      <td>TRAV5-4</td>\n",
       "      <td>TRAJ50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGGGAGTGCCATT-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACGGGAGTGCCATT-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>567</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV12N-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ31</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>...</td>\n",
       "      <td>spleen and lymph node</td>\n",
       "      <td>CD4+</td>\n",
       "      <td>5</td>\n",
       "      <td>CALNNNNRIFF</td>\n",
       "      <td>TRAV12N-2</td>\n",
       "      <td>TRAJ31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15482</th>\n",
       "      <td>TTTGTCACAAGTACCT-6</td>\n",
       "      <td>True</td>\n",
       "      <td>TTTGTCACAAGTACCT-6_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>595</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV5-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ34</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>...</td>\n",
       "      <td>spleen and lymph node</td>\n",
       "      <td>CD4+</td>\n",
       "      <td>15483</td>\n",
       "      <td>CAAGSSNTDKVVF</td>\n",
       "      <td>TRAV5-4</td>\n",
       "      <td>TRAJ34</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15483</th>\n",
       "      <td>TTTGTCATCAAACCAC-6</td>\n",
       "      <td>True</td>\n",
       "      <td>TTTGTCATCAAACCAC-6_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>580</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV12N-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ22</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>...</td>\n",
       "      <td>spleen and lymph node</td>\n",
       "      <td>CD4+</td>\n",
       "      <td>15484</td>\n",
       "      <td>CALSESSGSWQLIF</td>\n",
       "      <td>TRAV12N-2</td>\n",
       "      <td>TRAJ22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15484</th>\n",
       "      <td>TTTGTCATCACTCTTA-6</td>\n",
       "      <td>True</td>\n",
       "      <td>TTTGTCATCACTCTTA-6_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>599</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV5-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ50</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>...</td>\n",
       "      <td>spleen and lymph node</td>\n",
       "      <td>CD4+</td>\n",
       "      <td>15485</td>\n",
       "      <td>CAVSSSSFSKLVF</td>\n",
       "      <td>TRAV5-4</td>\n",
       "      <td>TRAJ50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15485</th>\n",
       "      <td>TTTGTCATCCGTAGTA-6</td>\n",
       "      <td>True</td>\n",
       "      <td>TTTGTCATCCGTAGTA-6_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>695</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-4</td>\n",
       "      <td>TRBC1</td>\n",
       "      <td>...</td>\n",
       "      <td>spleen and lymph node</td>\n",
       "      <td>CD4+</td>\n",
       "      <td>15486</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSMARGERLFF</td>\n",
       "      <td>TRBV19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15486</th>\n",
       "      <td>TTTGTCATCGGGAGTA-6</td>\n",
       "      <td>True</td>\n",
       "      <td>TTTGTCATCGGGAGTA-6_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>519</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV6N-7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ40</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>...</td>\n",
       "      <td>spleen and lymph node</td>\n",
       "      <td>CD4+</td>\n",
       "      <td>15487</td>\n",
       "      <td>CALSGLNTGNYKYVF</td>\n",
       "      <td>TRAV6N-7</td>\n",
       "      <td>TRAJ40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17425 rows × 54 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  barcode  is_cell                    contig_id  \\\n",
       "0      AAACCTGCAGCCACCA-1     True  AAACCTGCAGCCACCA-1_contig_1   \n",
       "1      AAACCTGCATTTCAGG-1     True  AAACCTGCATTTCAGG-1_contig_1   \n",
       "2      AAACCTGCATTTCAGG-1     True  AAACCTGCATTTCAGG-1_contig_2   \n",
       "3      AAACGGGAGGACTGGT-1     True  AAACGGGAGGACTGGT-1_contig_1   \n",
       "4      AAACGGGAGTGCCATT-1     True  AAACGGGAGTGCCATT-1_contig_1   \n",
       "...                   ...      ...                          ...   \n",
       "15482  TTTGTCACAAGTACCT-6     True  TTTGTCACAAGTACCT-6_contig_1   \n",
       "15483  TTTGTCATCAAACCAC-6     True  TTTGTCATCAAACCAC-6_contig_1   \n",
       "15484  TTTGTCATCACTCTTA-6     True  TTTGTCATCACTCTTA-6_contig_1   \n",
       "15485  TTTGTCATCCGTAGTA-6     True  TTTGTCATCCGTAGTA-6_contig_1   \n",
       "15486  TTTGTCATCGGGAGTA-6     True  TTTGTCATCGGGAGTA-6_contig_1   \n",
       "\n",
       "       high_confidence  length chain             v_gene d_gene   j_gene  \\\n",
       "0                 True     577   TRB  TRBV12-2+TRBV13-2    NaN  TRBJ2-7   \n",
       "1                 True     664   TRA            TRAV5-4    NaN   TRAJ42   \n",
       "2                 True     661   TRB           TRBV13-3    NaN  TRBJ2-7   \n",
       "3                 True     617   TRA            TRAV5-4    NaN   TRAJ50   \n",
       "4                 True     567   TRA          TRAV12N-2    NaN   TRAJ31   \n",
       "...                ...     ...   ...                ...    ...      ...   \n",
       "15482             True     595   TRA            TRAV5-4    NaN   TRAJ34   \n",
       "15483             True     580   TRA          TRAV12N-2    NaN   TRAJ22   \n",
       "15484             True     599   TRA            TRAV5-4    NaN   TRAJ50   \n",
       "15485             True     695   TRB             TRBV19    NaN  TRBJ1-4   \n",
       "15486             True     519   TRA           TRAV6N-7    NaN   TRAJ40   \n",
       "\n",
       "      c_gene  ...            meta.tissue  meta.cell.subset chunk.id  \\\n",
       "0      TRBC2  ...  spleen and lymph node              CD4+        1   \n",
       "1       TRAC  ...  spleen and lymph node              CD4+        2   \n",
       "2      TRBC2  ...  spleen and lymph node              CD4+        3   \n",
       "3       TRAC  ...  spleen and lymph node              CD4+        4   \n",
       "4       TRAC  ...  spleen and lymph node              CD4+        5   \n",
       "...      ...  ...                    ...               ...      ...   \n",
       "15482   TRAC  ...  spleen and lymph node              CD4+    15483   \n",
       "15483   TRAC  ...  spleen and lymph node              CD4+    15484   \n",
       "15484   TRAC  ...  spleen and lymph node              CD4+    15485   \n",
       "15485  TRBC1  ...  spleen and lymph node              CD4+    15486   \n",
       "15486   TRAC  ...  spleen and lymph node              CD4+    15487   \n",
       "\n",
       "            cdr3.alpha    v.alpha j.alpha          cdr3.beta  \\\n",
       "0                  NaN        NaN     NaN  CASGDALGLGVSYEQYF   \n",
       "1      CAASVAGGSNAKLTF    TRAV5-4  TRAJ42    CASSEAGTGGYEQYF   \n",
       "2      CAASVAGGSNAKLTF    TRAV5-4  TRAJ42    CASSEAGTGGYEQYF   \n",
       "3      CAASLASSSFSKLVF    TRAV5-4  TRAJ50                NaN   \n",
       "4          CALNNNNRIFF  TRAV12N-2  TRAJ31                NaN   \n",
       "...                ...        ...     ...                ...   \n",
       "15482    CAAGSSNTDKVVF    TRAV5-4  TRAJ34                NaN   \n",
       "15483   CALSESSGSWQLIF  TRAV12N-2  TRAJ22                NaN   \n",
       "15484    CAVSSSSFSKLVF    TRAV5-4  TRAJ50                NaN   \n",
       "15485              NaN        NaN     NaN      CASSMARGERLFF   \n",
       "15486  CALSGLNTGNYKYVF   TRAV6N-7  TRAJ40                NaN   \n",
       "\n",
       "                  v.beta d.beta   j.beta  \n",
       "0      TRBV12-2+TRBV13-2    NaN  TRBJ2-7  \n",
       "1               TRBV13-3    NaN  TRBJ2-7  \n",
       "2               TRBV13-3    NaN  TRBJ2-7  \n",
       "3                    NaN    NaN      NaN  \n",
       "4                    NaN    NaN      NaN  \n",
       "...                  ...    ...      ...  \n",
       "15482                NaN    NaN      NaN  \n",
       "15483                NaN    NaN      NaN  \n",
       "15484                NaN    NaN      NaN  \n",
       "15485             TRBV19    NaN  TRBJ1-4  \n",
       "15486                NaN    NaN      NaN  \n",
       "\n",
       "[17425 rows x 54 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['cdr3.alpha'] = data.barcode.apply(lambda x: barcodes_dict[x]['TRA'][3])\n",
    "data['v.alpha'] = data.barcode.apply(lambda x: barcodes_dict[x]['TRA'][0])\n",
    "data['j.alpha'] = data.barcode.apply(lambda x: barcodes_dict[x]['TRA'][2])\n",
    "\n",
    "\n",
    "data['cdr3.beta'] = data.barcode.apply(lambda x: barcodes_dict[x]['TRB'][3])\n",
    "data['v.beta'] = data.barcode.apply(lambda x: barcodes_dict[x]['TRB'][0])\n",
    "data['d.beta'] = data.barcode.apply(lambda x: barcodes_dict[x]['TRB'][1])\n",
    "data['j.beta'] = data.barcode.apply(lambda x: barcodes_dict[x]['TRB'][2])\n",
    "\n",
    "data = data.explode(['cdr3.beta', 'v.beta', 'd.beta', 'j.beta']).explode(['cdr3.alpha', 'v.alpha', 'j.alpha'])\n",
    "\n",
    "\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4275c034",
   "metadata": {},
   "outputs": [],
   "source": [
    "data.rename(columns = {\n",
    "    'donor': 'meta.subject.id',\n",
    "    'barcode': 'meta.clone.id'\n",
    "}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "08a546b1",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "Index(['barcode'], dtype='object')",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_57379/3073279246.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop_duplicates\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubset\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'barcode'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'cdr3.beta'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'v.beta'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'd.beta'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'j.beta'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'cdr3.alpha'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'v.alpha'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'j.alpha'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/py311/lib/python3.11/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, subset, keep, inplace, ignore_index)\u001b[0m\n\u001b[1;32m   6836\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6837\u001b[0m         \u001b[0minplace\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalidate_bool_kwarg\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minplace\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"inplace\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6838\u001b[0m         \u001b[0mignore_index\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalidate_bool_kwarg\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mignore_index\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"ignore_index\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6839\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6840\u001b[0;31m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mduplicated\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubset\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkeep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mkeep\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6841\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mignore_index\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6842\u001b[0m             \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdefault_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6843\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/py311/lib/python3.11/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, subset, keep)\u001b[0m\n\u001b[1;32m   6968\u001b[0m         \u001b[0;31m# Otherwise, raise a KeyError, same as if you try to __getitem__ with a\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6969\u001b[0m         \u001b[0;31m# key that doesn't exist.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6970\u001b[0m         \u001b[0mdiff\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubset\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6971\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mdiff\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6972\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mIndex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdiff\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6973\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6974\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubset\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m1\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_unique\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6975\u001b[0m             \u001b[0;31m# GH#45236 This is faster than get_group_index below\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: Index(['barcode'], dtype='object')"
     ]
    }
   ],
   "source": [
    "data = data.drop_duplicates(subset=['barcode', 'cdr3.beta', 'v.beta', 'd.beta', 'j.beta', 'cdr3.alpha', 'v.alpha', 'j.alpha'])\n",
    "\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78ef5629",
   "metadata": {},
   "outputs": [],
   "source": [
    "coulmns = ['chunk.id',\n",
    " 'cdr3.alpha',\n",
    " 'v.alpha',\n",
    " 'j.alpha',\n",
    " 'cdr3.beta',\n",
    " 'v.beta',\n",
    " 'd.beta',\n",
    " 'j.beta',\n",
    " 'species',\n",
    " 'mhc.a',\n",
    " 'mhc.b',\n",
    " 'mhc.class',\n",
    " 'antigen.epitope',\n",
    " 'antigen.gene',\n",
    " 'antigen.species',\n",
    " 'reference.id',\n",
    " 'method.identification',\n",
    " 'method.frequency',\n",
    " 'method.singlecell',\n",
    " 'method.sequencing',\n",
    " 'method.verification',\n",
    " 'meta.study.id',\n",
    " 'meta.cell.subset',\n",
    " 'meta.subset.frequency',\n",
    " 'meta.subject.cohort',\n",
    " 'meta.subject.id',\n",
    " 'meta.replica.id',\n",
    " 'meta.clone.id',\n",
    " 'meta.epitope.id',\n",
    " 'meta.tissue',\n",
    " 'meta.donor.MHC',\n",
    " 'meta.donor.MHC.method',\n",
    " 'meta.structure.id',\n",
    " 'comment']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29fff775",
   "metadata": {},
   "outputs": [],
   "source": [
    "for column in set(coulmns).difference(set(data.columns)):\n",
    "    data[column] = np.nan\n",
    "\n",
    "data = data[coulmns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4358af66",
   "metadata": {},
   "outputs": [],
   "source": [
    "data[data['antigen.epitope'].apply(lambda x: pd.notnull(x))].set_index('chunk.id').to_csv('PMID_41315082.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6806bc5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "data[data['antigen.epitope'].apply(lambda x: pd.isnull(x))].set_index('chunk.id').to_csv('PMID_41315082_negative.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7fbfc53",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py311",
   "language": "python",
   "name": "py311"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
